src/parse.cpp

```diff
@@ -431,6 +431,41 @@ expression* parser::parse_primary_expression(precedence prec) {
   case token_type::equal:
   case token_type::kw_in:
   case token_type::question: {
+    if (this->peek().type == token_type::star) {
+      lexer_transaction transaction = this->lexer_.begin_transaction();
+      token star_token = this->peek();
+      this->skip();
+      if (!this->peek().has_leading_newline) {
+        bool has_leading_async = this->peek().type == token_type::kw_async;
+        function_attributes attrb = function_attributes::generator;
+        if (has_leading_async) {
+          this->skip();
+          attrb = function_attributes::async_generator;
+        }
+        if (this->peek().type == token_type::kw_function) {
+          this->skip();
+          if (this->peek().type == token_type::identifier) {
+            this->error_reporter_->report(
+                error_generator_function_star_belongs_before_name{
+                    .function_name = this->peek().span(),
+                    .star = star_token.span()});
+          } else {
+            this->error_reporter_->report(
+                error_generator_function_star_belongs_after_keyword_function{
+                    .star = star_token.span()});
+          }
+          this->lexer_.roll_back_transaction(std::move(transaction));
+          this->skip();
+          if (has_leading_async) {
+            this->skip();
+          }
```

Comment:
These are the cases I see:

* `*` not followed by a function expression
* `*` followed by a newline
* `*async function f(){}`
* `*async function(){}`
* `*function f(){}`
* `*function(){}`

If we put lines 435-461 into a function, then the caller only cares about two cases:

* `*` not followed by a function expression, or `*` followed by a newline
* `*async function` or `*function`

For the former case, the caller doesn't need any extra information. For the latter case, the caller needs `attrb` and `star_token.begin` (see my other comment on line 463). `star_token.begin` can easily be determined by the caller (it can `this->peek()` before calling the helper function).

I think we could factor lines 435-461 into a function with the following signature:
```c++
std::optional<function_attributes> try_parse_function_with_leading_star();
```